Fill parameter values from environment variables added by the plugin

proski opened the pull request:
```
*  Fill parameter values from environment variables added by the plugin
   
   Parameters generate environment variables that take precedence over the
   environment variables supplied by the EnvironmentContributor extension.
   
   Some users relied on the plugin filling values for the parameters defined
   for the project. That functionality was broken in version 1.9.
   
   Make StashRepository#getParameters() fill parameters with the values that
   the plugin supplies to the environment.
   
   To avoid code duplication, move the environment variables to StashCause
   as a transient map (the data is already saved to XML files from the
   fields). Use that map both from StashBuildEnvironmentContributor and
   StashRepository.
   
   Make the environment variables override the parameters from the pull
   request comments. Users with comment posting permissions should not be
   able to override parameters that the administrator intended to be filled
   from the environment.
```

Reported by @Noltari. Someone else mentioned that earlier in a PR comment, but I cannot find that comment.

Noltari commented:
Thanks for looking into this @proski

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -437,4 +437,38 @@ public void startJobSkipsNullParameters() {
+
+    List<ParameterValue> parameters = captureBuildParameters();
+
+    assertThat(parameters, hasSize(1));
```
Please use hamcrest `contains()` it removes the need to a separate size check

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -437,4 +437,38 @@ public void startJobSkipsNullParameters() {
+    List<ParameterValue> parameters = captureBuildParameters();
+
+    assertThat(parameters, hasSize(1));
+    assertThat(parameters.get(0), instanceOf(StringParameterValue.class));
```
I'm not sure about the need to check the type here. What are the chances it will be something else while satisfying the other assertions?

proski replied:
Good point. Better yet, `contains()` takes care or it, as it requires constructing the expected value.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -437,4 +437,38 @@ public void startJobSkipsNullParameters() {
   }
+
+  @Test
+  public void parameters_populated_from_environment() {
```
calling it "...from_cause" would be more descriptive I think

proski replied:
`...from_StashCause` to be more precise.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -437,4 +437,38 @@ public void startJobSkipsNullParameters() {
+    assertThat(parameters, hasSize(1));
+    assertThat(parameters.get(0), instanceOf(StringParameterValue.class));
+    assertThat(parameters.get(0).getName(), is("sourceBranch"));
+    assertThat(parameters.get(0).getValue(), is("SourceBranch"));
```
It's a bit unclear where the value `SourceBranch` comes from. I think it's from line 95.
Maybe extracting that into a constant would make it easier for the readers?

proski replied:
You are right. I'm not sure how to do it better. Having variables for all arguments to the `StashCause` constructor would add a lot of code of little utility. Using a variable only for `sourceBranch` would be inconsistent and would suggest that we are treating it in a special way.

I can use names suggesting the branch, such as `source/branch` and `target/branch`, but again, I don't have a good idea how to do it consistently for arguments like `SourceRepositoryOwner`.

Leaving it as is for now.

jakub-bochenski replied:
Usually I would try to use a builder, changing the line `  cause = makeCause(prParameters);`
to something like this:
```
 cause = defaultCauseBuilder()
                .withAdditionalParameters(prParameters)
                .withSourceBranch("source branch from cause")
                .build();
```

jakub-bochenski replied:
I'm not insisting on change it now though, the code is still better than it was before

proski replied:
The builder pattern is not a good match for `StashCause`, as it has no reasonable default values for anything other than the additional parameters. The builder would not be useful in the production code.

I replaced `"SourceBranch"` with `cause.getSourceBranch()`, that should make the intention clear.

proski committed:
Fill parameter values from environment variables added by the plugin

Parameters generate environment variables that take precedence over the
environment variables supplied by the EnvironmentContributor extension.

Some users relied on the plugin filling values for the parameters defined
for the project. That functionality was broken in version 1.9.

Make StashRepository#getParameters() fill parameters with the values that
the plugin supplies to the environment.

To avoid code duplication, move the environment variables to StashCause
as a transient map (the data is already saved to XML files from the
fields). Use that map both from StashBuildEnvironmentContributor and
StashRepository.

Make the environment variables override the parameters from the pull
request comments. Users with comment posting permissions should not be
able to override parameters that the administrator intended to be filled
from the environment.

jakub-bochenski commented:
Please double-check I didn't mess anything when resolving conflicts

proski commented:
> Please double-check I didn't mess anything when resolving conflicts

It's good, thank you! I've rebased this PR on top of the master branch to keep the history more linear.

proski commented:
Make changes to `README.md`, it had no distinction between environment variables and parameters.
